pkg/upgrade/v1_17_0.go

```diff
@@ -2,20 +2,55 @@ package upgrade
 
 import (
 	"context"
+	"fmt"
 
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
 )
 
 func upgrade1_17_0(ctx context.Context, client client.Client, jaeger v1.Jaeger) (v1.Jaeger, error) {
-	d := []deprecationFlagMap{{
-		from: "collector.grpc.tls",
-		to:   "collector.grpc.tls.enabled",
-	}, {
-		from: "reporter.grpc.tls",
-		to:   "reporter.grpc.tls.enabled",
-	}}
-
-	return migrateAllDeprecatedOptions(jaeger, d), nil
+	prefix := []string{
+		"collector.grpc",
+		"reporter.grpc",
+		"es",
+		"es-archive",
+		"cassandra",
+		"cassandra-archive",
+		"kafka.consumer", // the option kafka.consumer.tls never really existed, but it's there now, and in deprecated mode...
+		"kafka.producer", // same as the above comment
+	}
+	d := []deprecationFlagMap{}
+
+	for _, item := range prefix {
+		d = append(d, deprecationFlagMap{
+			from: fmt.Sprintf("%s.tls", item),
+			to:   fmt.Sprintf("%s.tls.enabled", item),
+		})
+	}
+
+	jaeger = migrateAllDeprecatedOptions(jaeger, d)
+
+	// for the collector and ingester, if we have TLS options but not ".enabled", we add the ".enabled" option
+	collectorOpts := jaeger.Spec.Collector.Options.GenericMap()
+	if collectorOpts["kafka.producer.authentication"] == "tls" {
+		collectorOpts["kafka.producer.tls.enabled"] = "true"
+		jaeger.Spec.Collector.Options = v1.NewOptions(collectorOpts)
+	}
+
+	ingesterOpts := jaeger.Spec.Ingester.Options.GenericMap()
+	changed := false
+	if ingesterOpts["kafka.consumer.authentication"] == "tls" {
+		ingesterOpts["kafka.consumer.tls.enabled"] = "true"
+		changed = true
+	}
+	if ingesterOpts["kafka.producer.authentication"] == "tls" {
```

Comment:
Would the `kafka.producer...` options be defined in the ingester? Thought they would only be relevant in the collector options?

Reply:
When the ingester is configured with Kafka as the storage, it uses the producer configuration:

```console
$ podman run -e SPAN_STORAGE_TYPE=kafka jaegertracing/jaeger-ingester:1.17.0 --help | grep tls.enabled
2020/02/27 13:24:59 maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
      --kafka.consumer.tls                            (deprecated) see --kafka.consumer.tls.enabled
      --kafka.consumer.tls.enabled                    Enable TLS when talking to the remote server(s)
      --kafka.producer.tls                            (deprecated) see --kafka.producer.tls.enabled
      --kafka.producer.tls.enabled                    Enable TLS when talking to the remote server(s)
```